File: DailyChallenges/Scala/LeetCode_1802.scala
object Solution {
    def maxValue(n: Int, index: Int, maxSum: Int): Int = {
        def spread(l: Int, r: Int, c: Int, left: Int): Int = {
            val base = r - l + 1
            if (base == n) c + left / base // can spreed up once base = n
            else if (left < base) c 
            else spread(if (l > 0) l-1 else l, if (r < n-1) r+1 else r, c + 1, left - base)
        }
        
        spread(index, index, 1, maxSum - n)
    }
}

File: DailyChallenges/CPP/LeetCode_1802.cpp
class Solution {
public:
    int maxValue(int n, int index, int maxSum) {
        auto ms = [=](int x){
            int a1 = x-2, b1 = (a1<=index) ? 1 : (a1-index+1), c1 = (a1<=index)? a1 : index;
            int a2 = x-1, b2 = (a2<=n-index) ? 1 : (a2-n+index+1), c2 = (a2<=n-index) ? a2 : n-index;
            return (((long long)a1+b1)*c1 + ((long long)a2+b2)*c2)/2 + n; 
        };
        int l = 1, r = maxSum+1;
        while (l < r){
            int m = l + (r-l)/2;
            if (ms(m)<=maxSum) l = m+1; else r=m;
        } 
        return l-1;
    }
};
